app/Connect/Actions/BuildClientPatchDataAction.php

```diff
@@ -50,10 +51,24 @@ public function execute(
         }
 
         // If the hash is not marked as compatible, and the current user is not flagged to
-        // be testing the hash, return a dummy set that will inform the user.
-        if ($gameHash->compatibility !== GameHashCompatibility::Compatible
-                && $gameHash->compatibility_tester_id !== $user?->id) {
-            return $this->buildIncompatiblePatchData($game ?? $gameHash->game, $gameHash->compatibility, $user);
+        // be testing the hash, a QA member, or the set author, return a dummy set that will
+        // inform the user.
+        if ($gameHash->compatibility !== GameHashCompatibility::Compatible) {
+            $game ??= $gameHash->game;
+
+            $canSeeIncompatibleSet = false;
+            if ($user) {
+                if ($user->id === $gameHash->compatibility_tester_id
+                    || $user->hasRole(Role::QUALITY_ASSURANCE)) {
+                    $canSeeIncompatibleSet = true;
+                } else {
+                    $canSeeIncompatibleSet = $game->achievements()->where('user_id', $user->id)->exists();
+                }
+            }
```

Comment:
Recommend extracting this to a policy, maybe `GameHashPolicy`. This will likely have utility elsewhere in the future.

```php
// untested pseudocode
public function seeIncompatibleSet(User $user, GameHash $gameHash): bool
{
    // Compatibility testers can always access incompatible content.
    if ($user->is($gameHash->compatibilityTester)) {
        return true;
    }

    // QA members can always access incompatible content.
    if ($user->hasRole(Role::QUALITY_ASSURANCE)) {
        return true;
    }

    // Achievement authors can always access their own work.
    return $gameHash->game->achievements()->where('user_id', $user->id)->exists();
}
```